File: sdk/python/dango/ws.py
"""Long-lived, multiplexed client for the native `/ws` endpoint.

A single connection carries any number of subscriptions (`perpsEvents`,
`fullBlock`) *and* transaction broadcasts, demultiplexed by the protocol's
`id`/`channel`. This is the shared-connection counterpart to the one-shot
helpers it replaces: where the old per-operation client opened a fresh socket
for every subscription and every broadcast, `WsConnection` keeps one socket
open and routes each reply back to the caller that issued it — the pattern a
high-performance trading bot wants (subscribe, react, broadcast, all on one
connection).

The Rust SDK splits this into a `Clone` handle plus a private `WsManager`
actor joined by a channel; Python has no such ownership boundary, so the
manager is folded into this one class. A daemon reader thread routes inbound
frames into per-operation queues/slots shared under a lock, while
`subscribe`/`broadcast` run on the caller's thread (register the reply channel,
then send). A daemon keepalive thread sends an app-level ping every 20s so an
idle connection is not reaped by the server's 60s idle timeout.

Not to be confused with `WebsocketManager` (graphql-transport-ws over
`/graphql`, slated for deprecation).
"""

from __future__ import annotations

import contextlib
import json
import queue
import threading
from typing import Any, Final, cast

import websocket  # from websocket-client

from dango.utils.error import ServerError
from dango.utils.types import Tx

# App-level keepalive interval. The server closes a connection it has not heard
# from for 60s; a 20s ping keeps an idle connection alive with a comfortable
# margin (any inbound frame resets the server's idle timer).
_PING_INTERVAL_SECONDS: Final[float] = 20.0

# Sentinel enqueued into a subscription's queue when the socket closes, so the
# `Subscription` iterator ends cleanly (StopIteration) instead of blocking
# forever in `queue.get()`.
_CLOSED: Final[object] = object()


def _error_detail(error: Any) -> str:
    """Render a server `error` value (`{code, message}` or a scalar) as a string."""

    if isinstance(error, dict):
        return f"{error.get('code', 'error')}: {error.get('message', '')}"

    return str(error)


class _OneShot:
    """A broadcast reply slot: a `threading.Event` plus the parked result or error."""

    def __init__(self) -> None:
        self._event: threading.Event = threading.Event()
        self._result: dict[str, Any] | None = None
        self._error: str | None = None

    def set_ok(self, data: dict[str, Any]) -> None:
        self._result = data
        self._event.set()

    def set_err(self, message: str) -> None:
        self._error = message
        self._event.set()

    def wait(self, timeout: float) -> dict[str, Any]:
        if not self._event.wait(timeout):
            raise ServerError("timed out waiting for broadcast reply")

        if self._error is not None:
            raise ServerError(f"broadcast failed: {self._error}")

        return self._result or {}


class WsConnection:
    """One long-lived `/ws` socket multiplexing subscriptions and broadcasts."""

    @classmethod
    def connect(cls, base_url: str, *, timeout: float = 10.0) -> WsConnection:
        """Open the socket and start the reader + keepalive threads."""

        self = cls(base_url)

        try:
            ws = websocket.create_connection(self._ws_url, timeout=timeout)
        except Exception as exc:
            raise ServerError(f"WebSocket connection to {self._ws_url} failed: {exc}") from exc

        # `create_connection` leaves the connect timeout on the socket, which
        # would make the reader's `recv()` bail after `timeout` seconds of an
        # idle (but healthy) subscription. Clear it so `recv()` blocks until a
        # frame arrives; liveness is handled by the keepalive ping instead.
        with contextlib.suppress(Exception):
            ws.settimeout(None)

        self._ws = ws
        threading.Thread(target=self._run, name="dango-ws-reader", daemon=True).start()
        threading.Thread(target=self._keepalive, name="dango-ws-keepalive", daemon=True).start()

        return self

    def __init__(self, base_url: str) -> None:
        self._ws_url: str = _make_ws_url(base_url)
        self._ws: websocket.WebSocket | None = None

        # `_lock` guards the id allocator and both registries, touched by the
        # caller threads (subscribe/broadcast/unsubscribe) and the reader thread
        # (routing replies). `_send_lock` serialises writes to the socket across
        # the caller threads and the keepalive thread.
        self._lock: threading.Lock = threading.Lock()
        self._send_lock: threading.Lock = threading.Lock()
        self._next_id: int = 0
        self._subs: dict[int, queue.Queue[Any]] = {}
        self._pending: dict[int, _OneShot] = {}
        self._stop: threading.Event = threading.Event()

    # --- public API (runs on the caller's thread) ---------------------------

    def subscribe(self, subscription: dict[str, Any]) -> Subscription:
        """Subscribe with a raw `subscription` selector; return a `Subscription` iterator."""

        sub_id = self._alloc_id()
        q: queue.Queue[Any] = queue.Queue()
        with self._lock:
            self._subs[sub_id] = q

        self._send({"method": "subscribe", "id": sub_id, "subscription": subscription})

        return Subscription(self, sub_id, q)

    def subscribe_perps_events(
        self,
        *,
        since_block_height: int | None = None,
        event_types: list[str] | None = None,
        pair_ids: list[str] | None = None,
        users: list[str] | None = None,
        order_ids: list[str] | None = None,
        client_order_ids: list[str] | None = None,
    ) -> Subscription:
        """Subscribe to `perpsEvents`; each item is one block's `PerpsEvent2Batch`.

        The five filters AND together; `None` (or an omitted filter) does not
        filter on that field, while an empty list matches nothing. A
        `client_order_id` is unique only per sender, so combine
        `client_order_ids` with `users` to single out one trader's order.
        """

        subscription: dict[str, Any] = {"type": "perpsEvents"}
        if since_block_height is not None:
            subscription["since"] = since_block_height
        for key, values in (
            ("eventTypes", event_types),
            ("pairIds", pair_ids),
            ("users", users),
            ("orderIds", order_ids),
            ("clientOrderIds", client_order_ids),
        ):
            if values is not None:
                subscription[key] = values

        return self.subscribe(subscription)

    def broadcast(self, tx: Tx, *, timeout: float = 10.0) -> dict[str, Any]:
        """Broadcast a signed tx over the same socket; block for the `BroadcastTxOutcome`.

        A mempool-rejected tx returns normally (the rejection rides
        `check_tx.result`); a transport failure, a timeout, or a socket that
        closes before the reply raises `ServerError`.
        """

        req_id = self._alloc_id()
        slot = _OneShot()
        with self._lock:
            self._pending[req_id] = slot

        self._send({"method": "broadcast", "id": req_id, "tx": cast("dict[str, Any]", tx)})

        return slot.wait(timeout)

    def close(self) -> None:
        """Stop the threads, close the socket, and fail every outstanding operation."""

        if self._stop.is_set():
            return

        self._stop.set()
        if self._ws is not None:
            with contextlib.suppress(Exception):
                self._ws.close()

        self._fail_all("connection closed")

    def __enter__(self) -> WsConnection:
        return self

    def __exit__(self, *_exc: object) -> None:
        self.close()

    # --- reader + keepalive threads -----------------------------------------

    def _run(self) -> None:
        ws = self._ws
        assert ws is not None  # set before this thread is started

        while not self._stop.is_set():
            try:
                raw = ws.recv()
            except Exception:
                break

            if not raw:
                break

            # A single malformed frame should not tear down the connection.
            try:
                msg = json.loads(raw)
            except Exception:
                continue

            if isinstance(msg, dict):
                self._dispatch(msg)

        self._fail_all("connection closed")

    def _dispatch(self, msg: dict[str, Any]) -> None:
        channel = msg.get("channel")
        # `id` is absent on connection-level error frames; `dict.get`/`pop` with
        # `None` simply miss the (int-keyed) registries, so no guard is needed.
        msg_id: Any = msg.get("id")

        if channel == "broadcast":
            with self._lock:
                slot = self._pending.pop(msg_id, None)
            if slot is not None:
                # A transport failure to the consensus node is an `error` frame;
                # a mempool rejection arrives as a `data` frame instead.
                if "error" in msg:
                    slot.set_err(_error_detail(msg.get("error")))
                else:
                    slot.set_ok(cast("dict[str, Any]", msg.get("data", {}) or {}))
        elif channel in ("perpsEvents", "fullBlock"):
            # An `error` on a subscription channel (e.g. `resync`,
            # `tooManyRequests`) is terminal, so drop the registration; the
            # iterator surfaces it and ends.
            if "error" in msg:
                with self._lock:
                    q = self._subs.pop(msg_id, None)
            else:
                with self._lock:
                    q = self._subs.get(msg_id)
            if q is not None:
                q.put(msg)
        # `subscriptionResponse` / `pong` / a connection-level `error` (no id)
        # route to no single caller, so they are ignored.

    def _keepalive(self) -> None:
        while not self._stop.wait(_PING_INTERVAL_SECONDS):
            try:
                self._send({"method": "ping"})
            except Exception:
                break

    # --- internals ----------------------------------------------------------

    def _alloc_id(self) -> int:
        with self._lock:
            self._next_id += 1
            return self._next_id

    def _send(self, message: dict[str, Any]) -> None:
        ws = self._ws
        if ws is None:
            raise ServerError("WebSocket is not connected")

        with self._send_lock:
            ws.send(json.dumps(message))

    def _unsubscribe(self, sub_id: int) -> None:
        """Drop a subscription locally and tell the server to stop streaming."""

        with self._lock:
            existed = self._subs.pop(sub_id, None) is not None

        if existed and not self._stop.is_set():
            with contextlib.suppress(Exception):
                self._send({"method": "unsubscribe", "id": sub_id})

    def _fail_all(self, reason: str) -> None:
        with self._lock:
            subs = list(self._subs.values())
            self._subs.clear()
            pending = list(self._pending.values())
            self._pending.clear()

        for q in subs:
            q.put(_CLOSED)
        for slot in pending:
            slot.set_err(reason)


class Subscription:
    """A blocking iterator over one subscription's frames (the `SubscriptionStream` analog).

    Each `next()` blocks until the next `PerpsEvent2Batch` arrives and returns
    it. A terminal server `error` frame raises `ServerError`; a closed
    connection ends the iterator (`StopIteration`). Use it as a context manager
    (or call `unsubscribe()`) to tell the server to stop streaming.
    """

    def __init__(self, conn: WsConnection, sub_id: int, q: queue.Queue[Any]) -> None:
        self._conn: WsConnection = conn
        self._id: int = sub_id
        self._q: queue.Queue[Any] = q

    @property
    def id(self) -> int:
        return self._id

    def __iter__(self) -> Subscription:
        return self

    def __next__(self) -> dict[str, Any]:
        item = self._q.get()

        if item is _CLOSED:
            raise StopIteration

        return self._interpret(item)

    def next_batch(self, timeout: float | None = None) -> dict[str, Any]:
        """Return the next batch, blocking up to `timeout` seconds (None waits forever).

        Unlike iterating, this surfaces a stalled feed: it raises `TimeoutError`
        if no frame arrives in time (the iterator protocol has no way to signal
        that). A terminal server error raises `ServerError`; a closed connection
        raises `ServerError` too, since an explicit read cannot yield a clean end.
        """

        try:
            item = self._q.get(timeout=timeout)
        except queue.Empty as exc:
            raise TimeoutError(f"no perps event within {timeout}s") from exc

        if item is _CLOSED:
            raise ServerError("connection closed")

        return self._interpret(item)

    def _interpret(self, item: dict[str, Any]) -> dict[str, Any]:
        if "error" in item:
            raise ServerError(_error_detail(item.get("error")))

        # The data frame's payload is already the bare batch (no GraphQL `data`
        # wrapper); hand it straight to the caller.
        return cast("dict[str, Any]", item.get("data", {}) or {})

    def unsubscribe(self) -> None:
        self._conn._unsubscribe(self._id)

    def __enter__(self) -> Subscription:
        return self

    def __exit__(self, *_exc: object) -> None:
        self.unsubscribe()


def _make_ws_url(base_url: str) -> str:
    """Convert an http(s) base_url to the matching ws(s)://...ws endpoint."""

    if base_url.startswith("https://"):
        endpoint = "wss://" + base_url[len("https://") :]
    elif base_url.startswith("http://"):
        endpoint = "ws://" + base_url[len("http://") :]
    elif base_url.startswith(("wss://", "ws://")):
        endpoint = base_url
    else:
        raise ValueError(f"unsupported base_url scheme: {base_url!r}")

    endpoint = endpoint.rstrip("/")
    if not endpoint.endswith("/ws"):
        endpoint = endpoint + "/ws"

    return endpoint

File: dango/sdk/src/ws.rs
use {
    anyhow::{anyhow, bail},
    dango_primitives::{BroadcastTxOutcome, Tx},
    futures::{
        SinkExt, Stream, StreamExt,
        channel::{mpsc, oneshot},
        stream::{SplitSink, SplitStream},
    },
    reqwest::IntoUrl,
    serde::{Deserialize, de::DeserializeOwned},
    std::{
        collections::HashMap,
        marker::PhantomData,
        pin::Pin,
        sync::{
            Arc,
            atomic::{AtomicU64, Ordering},
        },
        task::{Context, Poll},
        time::Duration,
    },
    tokio_tungstenite::{MaybeTlsStream, WebSocketStream, connect_async, tungstenite::Message},
    url::Url,
};

/// App-level keepalive interval. The server closes a connection it has not heard
/// from for 60s; a 20s ping keeps an idle connection alive with a comfortable
/// margin (any inbound frame resets the server's idle timer).
const KEEP_ALIVE_INTERVAL: Duration = Duration::from_secs(20);

/// The WebSocket connection [`connect_async`] yields for a `ws(s)://` URL.
type WsConn = WebSocketStream<MaybeTlsStream<tokio::net::TcpStream>>;

/// Data frames delivered to a subscription: `Ok` carries the frame's `data`
/// payload, `Err` is a terminal error (the stream ends after it).
type FrameSender = mpsc::UnboundedSender<anyhow::Result<serde_json::Value>>;
type FrameReceiver = mpsc::UnboundedReceiver<anyhow::Result<serde_json::Value>>;

// ---- perps events data types ----

/// One block's matching perps-contract events, as delivered by the `/ws`
/// `perpsEvents` channel. Mirrors the `perps_events` payload shape.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PerpsEventsBatch {
    pub block_height: u64,
    /// Block timestamp, RFC 3339.
    pub created_at: String,
    pub events: Vec<PerpsEvent>,
}

/// A single perps-contract event within a [`PerpsEventsBatch`].
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PerpsEvent {
    pub idx: u32,
    pub event_type: String,
    pub user: Option<String>,
    pub pair_id: Option<String>,
    pub order_id: Option<String>,
    pub client_order_id: Option<String>,
    pub data: serde_json::Value,
}

// ---- the handle ----

/// A single, long-lived native-`/ws` connection that multiplexes any number of
/// subscriptions (`perpsEvents`) **and** transaction broadcasts over one socket,
/// demultiplexed by the protocol's `id`/`channel`.
///
/// This is the shared-connection path a latency-sensitive bot wants: subscribe,
/// react, and broadcast all ride the same socket, so a `broadcast` never opens a
/// second connection and the event feed keeps draining while a broadcast is in
/// flight. The one-shot `HttpClient` helpers (REST `broadcast_tx`, the GraphQL
/// subscriptions) remain the simple per-operation path.
///
/// `WsConnection` is a cheap-to-`Clone` handle (backed by an [`Arc`]); it owns
/// only the sending end of a command channel. The socket and the routing
/// registries live in a background actor task ([`WsManager`]), reached only by
/// message-passing, so there are no locks on the hot path. The socket closes
/// once the last `WsConnection` clone and every [`Subscription`] have dropped.
///
/// Not to be confused with the graphql-transport-ws [`crate::WsClient`] /
/// [`crate::Session`], which speak a different protocol against `/graphql`.
#[derive(Debug, Clone)]
pub struct WsConnection {
    inner: Arc<WsConnectionInner>,
}

#[derive(Debug)]
struct WsConnectionInner {
    commands: mpsc::UnboundedSender<Command>,
    next_id: AtomicU64,
}

impl Drop for WsConnectionInner {
    fn drop(&mut self) {
        let _ = self.commands.unbounded_send(Command::Close);
    }
}

enum Command {
    Subscribe {
        id: u64,
        message: serde_json::Value,
        tx: FrameSender,
    },
    Unsubscribe {
        id: u64,
    },
    Broadcast {
        id: u64,
        message: serde_json::Value,
        reply: oneshot::Sender<anyhow::Result<BroadcastTxOutcome>>,
    },
    Close,
}

impl WsConnection {
    /// Open a `/ws` connection and spawn its background actor.
    ///
    /// `url` is the HTTP base (e.g. `https://api.dango.zone`); the `ws(s)://…/ws`
    /// endpoint is derived from it. Native `/ws` has no handshake, so this just
    /// splits the socket and spawns the manager.
    pub async fn connect<U>(url: U) -> anyhow::Result<Self>
    where
        U: IntoUrl,
    {
        let ws_url = to_ws_url(url.into_url()?)?;

        let (ws, _response) = connect_async(ws_url.as_str())
            .await
            .map_err(|err| anyhow!("WebSocket connection failed: {err}"))?;

        let (sink, stream) = ws.split();
        let (commands, command_rx) = mpsc::unbounded::<Command>();

        tokio::spawn(
            WsManager {
                sink,
                stream,
                commands: command_rx,
                subs: HashMap::new(),
                pending: HashMap::new(),
            }
            .run(),
        );

        Ok(Self {
            inner: Arc::new(WsConnectionInner {
                commands,
                next_id: AtomicU64::new(1),
            }),
        })
    }

    /// Subscribe to perps-exchange events over the shared socket (`perpsEvents`
    /// channel). Yields one [`PerpsEventsBatch`] per block that has at least one
    /// matching event; a terminal `Err` (e.g. the server's `resync` /
    /// `tooManyRequests`) ends the stream.
    ///
    /// The five filters are sets that AND together; `None` (or an empty list)
    /// does not filter on that field. `since_block_height` replays the retained
    /// in-memory window from that height (inclusive) before the live tail.
    pub fn subscribe_perps_events(
        &self,
        since_block_height: Option<u64>,
        event_types: Option<Vec<String>>,
        pair_ids: Option<Vec<String>>,
        users: Option<Vec<String>>,
        order_ids: Option<Vec<String>>,
        client_order_ids: Option<Vec<String>>,
    ) -> Subscription<PerpsEventsBatch> {
        // Absent filters are omitted (match-all); present ones are sent as JSON
        // arrays. Mirrors the server's `perpsEvents` subscription selector.
        let mut subscription = serde_json::Map::new();
        subscription.insert("type".into(), "perpsEvents".into());

        if let Some(since) = since_block_height {
            subscription.insert("since".into(), since.into());
        }

        for (key, values) in [
            ("eventTypes", event_types),
            ("pairIds", pair_ids),
            ("users", users),
            ("orderIds", order_ids),
            ("clientOrderIds", client_order_ids),
        ] {
            if let Some(values) = values {
                subscription.insert(key.into(), values.into());
            }
        }

        self.subscribe(serde_json::Value::Object(subscription))
    }

    /// Broadcast a signed transaction over the shared socket (`broadcast`
    /// channel) and await its receipt — no second connection, and the event feed
    /// keeps draining while this awaits. Reply frames are correlated to requests
    /// by `id`, so several in-flight broadcasts never collide.
    ///
    /// As with REST, a mempool-rejected tx returns `Ok` (the rejection rides
    /// [`BroadcastTxOutcome::check_tx`]); only a transport failure is `Err`.
    pub async fn broadcast(&self, tx: Tx) -> anyhow::Result<BroadcastTxOutcome> {
        let id = self.inner.next_id.fetch_add(1, Ordering::Relaxed);
        let (reply, rx) = oneshot::channel();

        let message = serde_json::json!({
            "method": "broadcast",
            "id": id,
            "tx": serde_json::to_value(&tx)?,
        });

        self.inner
            .commands
            .unbounded_send(Command::Broadcast { id, message, reply })
            .map_err(|_| anyhow!("connection closed"))?;

        rx.await.map_err(|_| anyhow!("connection closed"))?
    }

    /// Allocate an id, register a frame channel, and send the subscribe command.
    /// Runs on the caller: no socket or registry access happens here.
    fn subscribe<T>(&self, subscription: serde_json::Value) -> Subscription<T>
    where
        T: DeserializeOwned,
    {
        let id = self.inner.next_id.fetch_add(1, Ordering::Relaxed);
        let (tx, rx) = mpsc::unbounded::<anyhow::Result<serde_json::Value>>();

        let message = serde_json::json!({
            "method": "subscribe",
            "id": id,
            "subscription": subscription,
        });

        let _ = self
            .inner
            .commands
            .unbounded_send(Command::Subscribe { id, message, tx });

        Subscription {
            rx,
            id: Some(id),
            commands: self.inner.commands.clone(),
            _conn: self.clone(),
            _phantom: PhantomData,
        }
    }
}

// ---- the subscription stream ----

/// A [`Stream`] of one subscription's frames (the Rust analog of the Python
/// SDK's `Subscription` iterator). Yields `Ok(T)` per data frame and a terminal
/// `Err` on a subscription error or a closed connection; dropping it sends an
/// `unsubscribe` for its id.
pub struct Subscription<T> {
    rx: FrameReceiver,
    id: Option<u64>,
    commands: mpsc::UnboundedSender<Command>,
    // Keeps the connection alive while the stream lives, mirroring the socket's
    // RAII close: the actor exits only once every handle and stream have dropped.
    _conn: WsConnection,
    _phantom: PhantomData<fn() -> T>,
}

impl<T> Stream for Subscription<T>
where
    T: DeserializeOwned + Unpin,
{
    type Item = anyhow::Result<T>;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        match Pin::new(&mut self.rx).poll_next(cx) {
            Poll::Ready(Some(Ok(value))) => {
                Poll::Ready(Some(serde_json::from_value::<T>(value).map_err(Into::into)))
            },
            Poll::Ready(Some(Err(err))) => Poll::Ready(Some(Err(err))),
            Poll::Ready(None) => Poll::Ready(None),
            Poll::Pending => Poll::Pending,
        }
    }
}

impl<T> Drop for Subscription<T> {
    fn drop(&mut self) {
        if let Some(id) = self.id.take() {
            let _ = self.commands.unbounded_send(Command::Unsubscribe { id });
        }
    }
}

// ---- the actor ----

/// The single owner of the socket and the routing registries. Reached only
/// through the command channel; never held by the caller. Runs on a spawned
/// task, so its writes and reads run concurrently with a comfortable keepalive.
struct WsManager {
    sink: SplitSink<WsConn, Message>,
    stream: SplitStream<WsConn>,
    commands: mpsc::UnboundedReceiver<Command>,
    /// Live subscriptions, by id → frame sender.
    subs: HashMap<u64, FrameSender>,
    /// In-flight broadcasts, by id → one-shot reply.
    pending: HashMap<u64, oneshot::Sender<anyhow::Result<BroadcastTxOutcome>>>,
}

impl WsManager {
    async fn run(self) {
        // Destructure into locals so the `select!` can borrow the socket ends,
        // the command channel, and the registries independently.
        let Self {
            mut sink,
            mut stream,
            mut commands,
            mut subs,
            mut pending,
        } = self;

        let mut keepalive = tokio::time::interval(KEEP_ALIVE_INTERVAL);
        // First tick is immediate; skip it so we don't ping right after connect.
        keepalive.tick().await;
        let ping = serde_json::json!({ "method": "ping" }).to_string();

        let reason = loop {
            tokio::select! {
                _ = keepalive.tick() => {
                    if sink.send(Message::text(ping.clone())).await.is_err() {
                        break "keepalive send failed";
                    }
                }
                cmd = commands.next() => match cmd {
                    Some(Command::Subscribe { id, message, tx }) => {
                        subs.insert(id, tx);
                        if sink.send(Message::text(message.to_string())).await.is_err() {
                            break "subscribe send failed";
                        }
                    }
                    Some(Command::Unsubscribe { id }) => {
                        subs.remove(&id);
                        let msg = serde_json::json!({ "method": "unsubscribe", "id": id });
                        if sink.send(Message::text(msg.to_string())).await.is_err() {
                            break "unsubscribe send failed";
                        }
                    }
                    Some(Command::Broadcast { id, message, reply }) => {
                        pending.insert(id, reply);
                        if sink.send(Message::text(message.to_string())).await.is_err() {
                            break "broadcast send failed";
                        }
                    }
                    // The last handle dropped (or an explicit close): shut down.
                    Some(Command::Close) | None => break "connection closed",
                },
                frame = stream.next() => match frame {
                    Some(Ok(Message::Text(text))) => dispatch(&mut subs, &mut pending, &text),
                    Some(Ok(Message::Ping(data))) => {
                        if sink.send(Message::Pong(data)).await.is_err() {
                            break "pong send failed";
                        }
                    }
                    Some(Ok(Message::Close(_))) | Some(Err(_)) | None => break "socket closed",
                    Some(Ok(_)) => {},
                },
            }
        };

        // Notify every outstanding subscription and broadcast so nobody hangs.
        for (_, tx) in subs.drain() {
            let _ = tx.unbounded_send(Err(anyhow!("{reason}")));
        }
        for (_, reply) in pending.drain() {
            let _ = reply.send(Err(anyhow!("{reason}")));
        }
        let _ = sink.close().await;
    }
}

/// Route one inbound text frame to the subscription or broadcast that owns its
/// `id`. An `error` key is co-located on the operation's own channel and is
/// terminal for that operation.
fn dispatch(
    subs: &mut HashMap<u64, FrameSender>,
    pending: &mut HashMap<u64, oneshot::Sender<anyhow::Result<BroadcastTxOutcome>>>,
    text: &str,
) {
    let Ok(value) = serde_json::from_str::<serde_json::Value>(text) else {
        return;
    };
    let id = value.get("id").and_then(serde_json::Value::as_u64);

    match value.get("channel").and_then(serde_json::Value::as_str) {
        // Mempool rejection is a `data` frame; only a transport failure to the
        // node is an `error` frame. Either way the broadcast is one-shot.
        Some("broadcast") => {
            if let Some(id) = id
                && let Some(reply) = pending.remove(&id)
            {
                let result = match value.get("error") {
                    Some(error) => Err(anyhow!("{}", error_detail(error))),
                    None => {
                        let data = value.get("data").cloned().unwrap_or_default();
                        serde_json::from_value::<BroadcastTxOutcome>(data).map_err(Into::into)
                    },
                };
                let _ = reply.send(result);
            }
        },
        Some("perpsEvents" | "fullBlock") => {
            if let Some(id) = id {
                match value.get("error") {
                    // Terminal: drop the registration and end the stream.
                    Some(error) => {
                        if let Some(tx) = subs.remove(&id) {
                            let _ = tx.unbounded_send(Err(anyhow!("{}", error_detail(error))));
                        }
                    },
                    None => {
                        if let Some(tx) = subs.get(&id) {
                            let data = value.get("data").cloned().unwrap_or_default();
                            let _ = tx.unbounded_send(Ok(data));
                        }
                    },
                }
            }
        },
        // `subscriptionResponse` / `pong` / a connection-level `error` (no id)
        // route to no single caller, so they are ignored.
        _ => {},
    }
}

/// Render a server `error` value (`{code, message}`) as `code: message`.
fn error_detail(error: &serde_json::Value) -> String {
    let code = error
        .get("code")
        .and_then(serde_json::Value::as_str)
        .unwrap_or("error");
    let message = error
        .get("message")
        .and_then(serde_json::Value::as_str)
        .unwrap_or_default();

    format!("{code}: {message}")
}

/// Derive the `ws(s)://…/ws` endpoint from an HTTP base URL.
fn to_ws_url(base: Url) -> anyhow::Result<Url> {
    let mut ws_url = base.join("ws")?;
    match ws_url.scheme() {
        "http" => ws_url
            .set_scheme("ws")
            .map_err(|_| anyhow!("failed to set ws scheme"))?,
        "https" => ws_url
            .set_scheme("wss")
            .map_err(|_| anyhow!("failed to set wss scheme"))?,
        "ws" | "wss" => {},
        scheme => bail!("invalid URL scheme: {scheme}"),
    }

    Ok(ws_url)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ws_url_derivation() {
        // http(s) map to ws(s), and the `/ws` path is appended to the base.
        for (base, expected) in [
            ("https://api.dango.zone", "wss://api.dango.zone/ws"),
            ("https://api.dango.zone/", "wss://api.dango.zone/ws"),
            ("http://localhost:8080", "ws://localhost:8080/ws"),
            ("ws://localhost:8080", "ws://localhost:8080/ws"),
        ] {
            let url = to_ws_url(Url::parse(base).unwrap()).unwrap();
            assert_eq!(url.as_str(), expected, "base={base}");
        }
    }

    #[test]
    fn ws_url_rejects_unknown_scheme() {
        assert!(to_ws_url(Url::parse("ftp://example.com").unwrap()).is_err());
    }
}
